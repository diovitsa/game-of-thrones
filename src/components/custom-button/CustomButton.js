import React from 'react';
import { TouchableOpacity, StyleSheet, Text } from 'react-native';


export const CustomButton = ({ title = 'Enter', style = {}, textStyle = {}, onPress, disabled }) => {

  return (
    <TouchableOpacity disabled={disabled} onPress={onPress}
                      style={[styles.button, style, disabled ? styles.disabled : null]}>
      <Text style={[styles.text, textStyle]}>{title}</Text>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    display: 'flex',
    height: 30,
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    margin: 5,
    padding: 5,

    backgroundColor: '#2AC062',
    shadowColor: '#2AC062',
    shadowOpacity: 0.4,
    shadowOffset: { height: 10, width: 0 },
    shadowRadius: 20,
  },

  text: {
    fontSize: 14,
    textTransform: 'uppercase',
    color: '#FFFFFF',
  },
  disabled: {
    backgroundColor: '#D3D3D3',
    shadowColor: '#D3D3D3',
  }
});