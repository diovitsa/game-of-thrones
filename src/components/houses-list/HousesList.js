import React, { useEffect } from 'react';
import { ScrollView, View, StyleSheet } from 'react-native';
import { connect } from 'react-redux';
import { fetchHousesData, nextPage, prevPage } from '../../actions/ac';
import House from '../house/House';
import { articlesLoadingSelector, housesSelector } from '../../selectors';
import LoadingBar from '../loading-bar/LoadingBar';
import { CustomButton } from '../custom-button/CustomButton';

function HousesList({ fetchHousesData, houses, page, nextPage, prevPage, isLoading }) {
  useEffect(() => {
    fetchHousesData(page);
  }, [page]);

  const housesList = houses
    ? houses.data.map((house, index) => {
      return <House house={house} id={index} key={index}/>
    })
    : null;

  function handleButtonClick(isNext) {
    return isNext
      ? nextPage()
      : prevPage();
  }

  return (
    <ScrollView>
      {isLoading ? <LoadingBar/> : <View>{housesList}</View>}
      <View style={styles.flex}>
        <CustomButton disabled={page === 1} onPress={() => handleButtonClick(false)} title={'Prev Page'}/>
        <CustomButton onPress={() => handleButtonClick(true)} title={'Next Page'}/>
      </View>
    </ScrollView>
  );
}

const mapDispatchToProps = {
  fetchHousesData,
  nextPage,
  prevPage
};

const mapStateToProps = state => ({
  houses: housesSelector(state),
  page: state.page,
  isLoading: articlesLoadingSelector(state)
});

const styles = StyleSheet.create({
  flex: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row'
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(HousesList);