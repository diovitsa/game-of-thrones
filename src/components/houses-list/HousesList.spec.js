import React from 'react';
import { HousesList } from './HousesList';
import renderer, { act } from 'react-test-renderer'
import { ScrollView } from 'react-native';
import { CustomButton } from '../custom-button/CustomButton';
import { head } from 'lodash';

describe('HousesList', () => {
  let component;
  let wrapper;

  const emptyData = {
    houses: null,
    fetchHousesData: jest.fn(),
    page: 1,
    nextPage: jest.fn(),
    prevPage: jest.fn(),
    isLoading: true
  };

  const loadedData = {
    ...emptyData,
    houses: { data: [{ name: 'testName1' }, { name: 'testName2' }] },
    isLoading: false
  };

  describe('when data still loading', () => {
    it('should render properly', () => {
      component = renderer.create(<HousesList {...emptyData} />);
      wrapper = component.root.findByType(ScrollView);

      expect(wrapper.props.children).toMatchSnapshot();
    });
  });

  describe('on data loaded', () => {
    it('should render properly', () => {
      component = renderer.create(<HousesList {...loadedData} />);
      wrapper = component.root.findByType(ScrollView);

      expect(wrapper.props.children).toMatchSnapshot();
    });
  });

  describe('buttons', () => {
    let prevBtn;
    let nextBtn;

    beforeEach(() => {
      component = renderer.create(<HousesList {...loadedData} />);
      const buttons = component.root.findAllByType(CustomButton);

      prevBtn = head(buttons.filter(button => button.props.id === 'prevPageBtn'));
      nextBtn = head(buttons.filter(button => button.props.id === 'nextPageBtn'));
    });

    it('should change pages', () => {
      act(() => {
        prevBtn.props.onPress();
      });

      expect(component.root.props.prevPage).toHaveBeenCalled();

      act(() => {
        nextBtn.props.onPress();
      });

      expect(component.root.props.nextPage).toHaveBeenCalled();

    });
  });
});