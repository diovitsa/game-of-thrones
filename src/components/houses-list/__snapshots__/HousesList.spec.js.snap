// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HousesList on data loaded should render properly 1`] = `
Array [
  <View>
    <House
      house={
        Object {
          "name": "testName1",
        }
      }
      id={0}
    />
    <House
      house={
        Object {
          "name": "testName2",
        }
      }
      id={1}
    />
  </View>,
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <CustomButton
      disabled={true}
      id="prevPageBtn"
      onPress={[Function]}
      title="Prev Page"
    />
    <CustomButton
      id="nextPageBtn"
      onPress={[Function]}
      title="Next Page"
    />
  </View>,
]
`;

exports[`HousesList when data still loading should render properly 1`] = `
Array [
  <LoadingBar />,
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <CustomButton
      disabled={true}
      id="prevPageBtn"
      onPress={[Function]}
      title="Prev Page"
    />
    <CustomButton
      id="nextPageBtn"
      onPress={[Function]}
      title="Next Page"
    />
  </View>,
]
`;
